File: cpp/121.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int maxProfit(vector<int> &prices) {
    int ans = 0;
    if(prices.empty()) return 0;
    int mi = prices[0];
    for(uint i=1; i<prices.size(); ++i){
      ans = max(ans, prices[i] - mi);
      mi = min(mi, prices[i]);
    }
    return ans;
  }
};

int main(){
  auto sol = Solution();
}

File: python/121.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        mi = prices[0]
        ans = 0
        for p in prices:
            ans = max(ans, p - mi)
            mi = min(mi, p)
        return ans
